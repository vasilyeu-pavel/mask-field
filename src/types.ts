import { ChangeEvent, InputHTMLAttributes, MutableRefObject, ReactNode } from "react"

export const typeCasting = <T>(data: unknown) => data as T

interface IInput extends Omit<InputHTMLAttributes<HTMLInputElement>, "className"> {
    errors?: string[] | null;
    modifiers?: string;
}

export type TSetCursorPosition = (value: number) => void
export type TSeparator = string
export type TSeparators = TSeparator[]

export interface IChildrenProps {
    maskInputStyle: string;
    maskInputWrapperStyle: string;
    handleMaskChange: (e: ChangeEvent<HTMLInputElement>) => void;
    inputRef: MutableRefObject<HTMLInputElement>;
    value: string;
}

export type TValidator = (value: string) => string | undefined

export interface IMaskWrapperProps extends Omit<IInput, "onChange" | "value"> {
    mask: string;
    value: string;
    children: (props: IChildrenProps) => ReactNode;
    separators: TSeparators;
    onChange: (maskedValue: string, originalValue: string) => void;
    validators?: TValidator[];
    handleErrors?: (errors?: string[]) => void;
}

export interface IMaskInputProps extends
    Omit<IMaskWrapperProps, "children">
{
    type?: string;
    withErrors?: boolean;
    modifiersErrors?: string;
}
