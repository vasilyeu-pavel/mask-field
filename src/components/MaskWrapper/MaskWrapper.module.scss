.maskContainer {
  --input-font-size: 12px;
  --default-ff: 'Montserrat', sans-serif;
  --background-jumper: #232f4d;
  --input-placeholder-color: #b9b9b9;

  position: relative;
  display: flex !important;
  align-items: center;
  cursor: text;
  border: 1px solid black;
  --mask-font: var(--input-font-size)/1 var(--default-ff);
}

.focus {
  border-color: var(--background-jumper);
}

.maskInput {
  z-index: 1;
  width: var(--mask-input-width);
  height: 100%;
  padding: 0;
  font: var(--mask-font);
  background: inherit;
  border: 0;
  border-radius: 0;
  box-shadow: none;
  &:focus {
    border: 0;
    outline: 0;
  }
}

.maskInputWrapper {
  position: relative;
  height: 100%;
}

.fakeSpan {
  position: absolute;
  z-index: 0;
  font: var(--mask-font);
  opacity: 0;
}

.mask {
  font: var(--mask-font);
  color: var(--input-placeholder-color);
}
