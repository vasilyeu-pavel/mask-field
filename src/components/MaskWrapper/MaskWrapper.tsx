import React, {
  ChangeEvent,
  memo,
  useCallback,
  useEffect,
  useRef,
  useState,
} from "react"
import { MASK_INPUT_WIDTH, MIN_DELAY, OFFSETS } from "../../constant"
import useCursorPosition from "../../hooks/useCursorPosition"
import { IMaskWrapperProps } from "../../types"
import {
  getMaskValue,
  handleMaskClick,
  isNumber,
  joinClasses,
  prepareValueFromMask,
  prepareValueToMask,
  setCSSVar,
} from "../../utils"

import style from "./MaskWrapper.module.scss"

export const MaskWrapper = (
  {
    mask,
    placeholder,
    children,
    value,
    errors,
    separators,
    onChange,
    validators = [isNumber],
    modifiers,
    handleErrors,
    readOnly,
    disabled,
  }: IMaskWrapperProps,
) => {
  const [maskValue, setMaskValue] = useState(mask)
  const [isFocus, setIsFocus] = useState(false)
  const [cursorPosition, setCursorPosition] = useState(0)
  const [valueToMaskStyle, setValueToMaskStyle] = useState("")
  const [pastedValue, setPastedValue] = useState("")

  const fakeInputRef = useRef < HTMLSpanElement | null >(null)
  const labelRef = useRef < HTMLLabelElement | null >(null)

  const handleValueToMaskStyle = useCallback((toInsertValue: string) => {
    const { maskedValue } = prepareValueToMask(toInsertValue, mask, separators)

    setValueToMaskStyle(maskedValue)
  }, [mask, separators, setValueToMaskStyle])

  /**
     * effects which came input value and prepare to mask style
     */
  useEffect(() => {
    handleValueToMaskStyle(value)
  }, [value, handleValueToMaskStyle])

  /**
     * listens mask changes, if mask was changed move cursor to end input
     */
  useEffect(() => {
    const { maskedValue } = prepareValueToMask(valueToMaskStyle, mask, separators)

    maskedValue.length && onChangeWrapper(maskedValue)
    moveCursorToEnd(maskedValue)
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [mask])

  /**
     * sets css variables which writes width to visible input
     * to continue offsets span-mask by n symbol
     * widths (offsetWidth) are gets on hidden span
     * IMPORTANT: sets css variables if all fonts was loaded
     */
  useEffect(() => {
    if (!fakeInputRef ||
            !(fakeInputRef && fakeInputRef.current) ||
            !labelRef ||
            !(labelRef && labelRef.current)
    ) return

    const fakeInput = fakeInputRef.current
    const label = labelRef.current

    setCSSVar(label, MASK_INPUT_WIDTH, "1px")

    // should set css variables if only all fonts was loaded
    document.fonts.ready.then(() => {
      const width = fakeInput.offsetWidth ? fakeInput.offsetWidth + OFFSETS : 1

      setCSSVar(label, MASK_INPUT_WIDTH, `${width}px`)
    })
  }, [maskValue, valueToMaskStyle])

  /**
     * effect which slices mask
     * if value 0 should be set mask
     */
  useEffect(() => {
    if (valueToMaskStyle) {
      setMaskValue(mask.slice(valueToMaskStyle.length))
    } else {
      setMaskValue(mask)
    }
  }, [valueToMaskStyle, mask])

  /**
     * set focus
     * * props: <focus: true/false>
     */
  const handleMaskFocus = useCallback((focus: boolean) => setIsFocus(focus), [setIsFocus])

  /**
     * move cursor to end input position
     * props: <value: string>
     */
  const moveCursorToEnd = useCallback(
    (value: string) => setCursorPosition(value.length),
    [setCursorPosition],
  )

  const errorsCB = useCallback(errors => {
    handleErrors && handleErrors(errors)

    moveCursorToEnd(valueToMaskStyle)
  }, [handleErrors, moveCursorToEnd, valueToMaskStyle])

  const onChangeWrapper = useCallback((
    insertValue: string,
    selectionStart?: number | null,
  ) => {
    !!pastedValue && setPastedValue("") // reset pasted value if pastedValue is not empty
    handleErrors && handleErrors() // reset errors

    const maskedOption = getMaskValue({
      value: insertValue,
      selectionStart: selectionStart || 0,
      mask,
      separators,
      validators,
      prevMaskValue: valueToMaskStyle,
      cursorPosition,
      errorsCB,
      pastedValue,
    })

    if (!maskedOption) return

    const { toMoveCursor, toInsertValue } = maskedOption

    setCursorPosition(toMoveCursor)
    handleValueToMaskStyle(toInsertValue)

    onChange(toInsertValue, prepareValueFromMask(toInsertValue, separators))
  }, [
    separators,
    onChange,
    mask,
    cursorPosition,
    errorsCB,
    handleValueToMaskStyle,
    validators,
    valueToMaskStyle,
    pastedValue,
    setPastedValue,
    handleErrors,
  ])

  /**
     * wrapper in  input.onChange
     *  runs validation and prepare entered value to mask style, set cursor
     *  returns:
     *  event - input event in which value was prepared to mask style ("12/12/")
     *  originalValue - value was not prepared to mask styled (1212)
     * @param e
     */
  const handleMaskChange = useCallback((e: ChangeEvent<HTMLInputElement>) => {
    onChangeWrapper(e.target.value, e.target.selectionStart)
  }, [onChangeWrapper])

  const isTouch = "ontouchend" in window

  const clickHandler = isTouch ? {
    /**
         * setTimeout needs because touchEvent not return correct selectionStart(cursorPosition)
         * when debugging was found that he returns events but later sets new cursorPosition
         * clickEvent doesn't have this problem
         */
    onTouchEnd: ({ target }: React.SyntheticEvent<HTMLLabelElement, TouchEvent>) => setTimeout(() =>
      handleMaskClick({
        valueToMaskStyle,
        setCursorPosition,
        handleMaskFocus,
        isTouch,
      })(target as HTMLInputElement),
    MIN_DELAY,
    ),
  } : {
    onClick: ({ target }: React.SyntheticEvent<HTMLLabelElement, MouseEvent>) =>
      handleMaskClick({
        valueToMaskStyle,
        setCursorPosition,
        handleMaskFocus,
        isTouch,
      })(target as HTMLInputElement),
  }

  return (
    <label
      className={joinClasses(
        style.maskContainer,
        "mask",
        readOnly && "readOnly",
        disabled && "disabled",
        isFocus && "focus",
        errors && errors.length && "wrong",
        modifiers,
      )}
      ref={labelRef}
      {...clickHandler}
    >
      <span ref={fakeInputRef} className={joinClasses(style.fakeSpan, "mask__fake")}>
        {valueToMaskStyle}
      </span>
      {(() => {
        // eslint-disable-next-line react-hooks/rules-of-hooks
        const inputRef = useCursorPosition<HTMLInputElement>({
          setCursorPosition,
          cursorPosition,
          isFocus,
          setPastedValue,
          handleMaskFocus,
        })

        return children({
          handleMaskChange,
          inputRef,
          maskInputWrapperStyle: style.maskInputWrapper,
          maskInputStyle: joinClasses(style.maskInput, "mask__input"),
          value: valueToMaskStyle,
        })
      })()}
      <span
        className={joinClasses(style.mask, "mask__tmp")}
        dangerouslySetInnerHTML={{ __html: (() => {
          if (!placeholder) return maskValue

          if (placeholder && value.length) {
            return maskValue
          } else {
            return placeholder
          }
        })().replace(/ /g, "&nbsp;"), // заменить все пробелы на html пробел
        }}
      />
    </label>
  )
}

export default memo(MaskWrapper)
