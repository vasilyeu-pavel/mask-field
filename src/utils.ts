import {
  CHAR_SEPARATOR_TEMPLATE,
  CHAR_VALUE_TEMPLATE,
  INPUT,
  LEFT,
  MIN_MAX_LENGTH,
  PRESS_KEY_EVENTS,
  RIGHT,
} from "./constant"
import {
  TSeparators,
  TSetCursorPosition,
  TValidator,
} from "./types"

export const makeHandleKeyDown = (setCursorPosition: (value: number) => void) =>
  (e: KeyboardEvent) => {
    if (!PRESS_KEY_EVENTS.includes(e.key)) return

    e.preventDefault()

    const { selectionStart } = e.target as HTMLInputElement
    const cursorPosition = selectionStart || 0

    switch (e.key) {
      case LEFT: {
        if (cursorPosition - 1 < 0) return

        return setCursorPosition(cursorPosition - 1)
      }
      // eslint-disable-next-line no-fallthrough
      case RIGHT: return setCursorPosition(cursorPosition + 1)
    }
  }

export const isNumber = (value: string): string | undefined => {
  if (!/^\d+$/.test(value)) {
    return "Value is not a number"
  }
}

export const getMaskOptions = (
  mask: string,
  separators: string[],
  startPosition = 0,
  finishedPosition?: number,
) => {
  const result: {
    separators: {
      separator: string,
      position: number,
    }[],
    template: number[],
  } = {
    separators: [],
    template: [],
  }

  for (let i = startPosition; i < (finishedPosition || mask.length); i++) { // eslint-disable-line no-plusplus,max-len
    const char = mask[i]

    if (separators.includes(char)) {
      result.separators.push({ separator: char, position: i })
      result.template.push(CHAR_SEPARATOR_TEMPLATE)
    } else {
      result.template.push(CHAR_VALUE_TEMPLATE)
    }
  }

  return result
}

/**
 * prepares mask value to origin view
 * "12/32/12" -> "123212"
 * @param value
 */
export const prepareValueFromMask = (value: string, separators: string[]): string => {
  let preparedValue = value

  separators.forEach(separator => {
    preparedValue = preparedValue.replace(new RegExp(`\\${separator}`, "g"), "")
  })

  return preparedValue
}

/**
 * prepares value to mask style
 * маска "__/__/__"
 * value "123" -> "12/3"
 * @param value
 */
export const prepareValueToMask = (
  value: string,
  mask: string,
  separators: string[],
): {
  maskedValue: string,
  originalValue: string,
} => {
  const originalValue = prepareValueFromMask(value, separators)
  const indexes = getMaskOptions(mask, separators).separators
  const arr = originalValue.split("")

  indexes.forEach(({ position, separator }) => {
    if (position < arr.length) {
      arr.splice(position, 0, separator)
    }
  })

  return {
    maskedValue: arr.join(""),
    originalValue,
  }
}

/**
 * var for saved cursorPosition when was called click on label/span
 */
let tmp: number | null = null

/**
 * saves current cursor position on click by input field
 */
export const handleMaskClick = (
  {
    valueToMaskStyle,
    setCursorPosition,
    handleMaskFocus,
    isTouch,
  }: {
      valueToMaskStyle: string,
      setCursorPosition: TSetCursorPosition,
      handleMaskFocus: (isFocus: boolean) => void,
      isTouch: boolean,
    },
) => (target: HTMLInputElement | null) => {
  if (!target) return

  // set focus
  // needs for firefox when we set cursor ff trying to highlight all input text
  target.focus()

  if (target.tagName !== INPUT) {
    // saving where move cursor after click on input
    tmp = valueToMaskStyle.length

    // on touch events aren't skipping because they didn't call two times
    if (!isTouch) return
  }

  // if we saved tmp position we should call with this value
  // another case we can use current cursor position from input (selectionStart)
  const value = tmp || target.selectionStart || 0

  setCursorPosition(value)

  handleMaskFocus(true)
  // reset tmp
  tmp = null
}

export /**
 * MASK MAIN METHOD (CORE)
 * moves cursor
 * returns value in mask style
 */
const getCursorPosition = (
  {
    nextMaskValue,
    toMove,
    prevMaskValue,
    separators,
    cursorPosition,
    toMoveCursorAfterPasted,
    mask,
  }: {
      nextMaskValue: string,
      toMove: number, // default from input
      prevMaskValue: string,
      separators: TSeparators,
      cursorPosition: number,
      toMoveCursorAfterPasted: number,
      mask: string,
    },
): {
  toMoveCursor: number,
  toInsertValue: string,
} => {
  const diff = nextMaskValue.length - prevMaskValue.length

  const originalDiff = (prepareValueFromMask(nextMaskValue, separators).length -
      prepareValueFromMask(prevMaskValue, separators).length)

  let toMoveCursor = nextMaskValue.length

  const {
    plus,
    minus,
  } = getSeparatorsArroundCursor(mask, separators, cursorPosition)

  // paste mode
  if (toMoveCursorAfterPasted) {
    // was added to cursor position available offsets for sliced pasted string
    // empty field -> paste 123 -> 12/3
    // (toMoveCursorAfterPasted = 4 about 3 digits from value and one of separator)

    return {
      toMoveCursor: cursorPosition + toMoveCursorAfterPasted,
      toInsertValue: nextMaskValue,
    }
  }

  // remove all separators (left) (backspace)
  if (cursorPosition > toMove && separators.includes(nextMaskValue[toMove])) {
    const removedSymbols = 1

    // 12/|2 => 1|2
    // 12/| => 1|
    toMoveCursor = cursorPosition - minus - removedSymbols

    return {
      toMoveCursor,
      // was removed prev symbol if you touch backspace after separator // 123 |41 -> 12|4 1
      toInsertValue: nextMaskValue
        .split("")
        .filter((a, i) => !!(i !== toMoveCursor))
        .join(""),
    }
  }

  // remove one symbol left
  // 123| => 12| (diff=1)
  // 12/3| => 12| (diff=2)
  if (diff < 0) {
    const isRemoveToLeft = cursorPosition > toMove
    const removedChar = 1
    const prevChar = 1
    let toMoveCursor = toMove

    if (isRemoveToLeft && separators.includes(mask[cursorPosition - removedChar - prevChar])) {
      const {
        minus,
      } = getSeparatorsArroundCursor(mask, separators, cursorPosition - prevChar)

      toMoveCursor = toMove - minus
    }

    return {
      toMoveCursor,
      toInsertValue: nextMaskValue,
    }
  }

  // remove all separator by delete btn (right)
  // 12|/45 => 12/|5
  if (toMove === cursorPosition && separators.includes(nextMaskValue[cursorPosition])) {
    toMoveCursor = cursorPosition + plus

    return {
      // 123| 3 -> 123 | -> 123|
      // toMoveCursor - 4 toInsertValue - 3(123) -> should set toInsertValue.length
      toMoveCursor: toMoveCursor > nextMaskValue.length ? nextMaskValue.length : toMoveCursor,
      // was removed next symbol if you touch delete before separator // 123| 3 -> 123 | -> 123|
      // and should prepare new input value to mask style (remove separators)
      toInsertValue: prepareValueToMask(
        nextMaskValue
          .split("")
          .filter((a, i) => !!(i !== toMoveCursor))
          .join(""),
        mask,
        separators,
      ).maskedValue,
    }
  }

  // add
  if (diff > 0) {
    // добавить диф и сместить на кол-во сепараторов если след. сепаратор

    return {
      toMoveCursor: cursorPosition + plus + originalDiff,
      toInsertValue: nextMaskValue,
    }
  }

  // was called when input was full
  if (diff === 0 && toMoveCursor !== cursorPosition) {
    // 1|2/12/1992 => 13|/21/2199
    toMoveCursor = toMove

    if (separators.includes(nextMaskValue[cursorPosition])) {
      // 12|/12/1992 => 12/31/2199
      toMoveCursor = toMove + plus
    }

    return {
      toMoveCursor,
      toInsertValue: nextMaskValue,
    }
  }

  return {
    toMoveCursor,
    toInsertValue: nextMaskValue,
  }
}

/**
 * validators
 * comes value and array validators and returns Array error or undefined
 * if not one of each validator not found error
 * @param value
 */
export const handleValidators = (value: string) =>
  (validators: TValidator[]): Array<string> | undefined => {
    const errorsFromValidators = validators.reduce((
      acc: Array<string>,
      validator: TValidator,
    ) => {
      const error = validator(value)

      value.length && error && acc.push(error)

      return acc
    }, [])

    if (!errorsFromValidators.length) return

    return errorsFromValidators
  }

export const getSeparatorsArroundCursor = (
  mask: string,
  separators: TSeparators,
  cursorPosition: number,
): {
  minus: number,
  plus: number,
} => {
  const template = getMaskOptions(mask, separators).template

  const counts = {
    plus: 0,
    minus: 0,
  }

  let isAvailablePlus = true
  let isAvailableMinus = true

  for (let i = cursorPosition; i < template.length; i++) { // eslint-disable-line no-plusplus
    if (template[i] === CHAR_SEPARATOR_TEMPLATE && isAvailablePlus) {
      counts.plus = counts.plus + 1
    }

    if (template[i] === CHAR_VALUE_TEMPLATE && isAvailablePlus) {
      isAvailablePlus = false
    }
  }

  for (let i = cursorPosition - 1; i >= 1; i--) { // eslint-disable-line no-plusplus
    if (template[i] === CHAR_SEPARATOR_TEMPLATE && isAvailableMinus) {
      counts.minus = counts.minus + 1
    }

    if (template[i] === CHAR_VALUE_TEMPLATE && isAvailableMinus) {
      isAvailableMinus = false
    }
  }

  return counts
}

/**
 * return real counts char and separators
 * for example: you want paste 444 to mask xx/xx/xxxx
 * should return chars: 2, separators: 1
 */
export const getCharsInfo = (
  mask: string,
  separators: TSeparators,
  cursorPosition: number,
  pastedValue: string,
): {
  chars: number,
  separators: number,
} => {
  const template = getMaskOptions(mask, separators).template

  const counts = {
    chars: 0,
    separators: 0,
  }

  for (let i = cursorPosition; i < template.length; i++) { // eslint-disable-line no-plusplus
    if (template[i] === CHAR_VALUE_TEMPLATE && pastedValue.length > counts.chars) {
      counts.chars = counts.chars + 1
    }

    if (template[i] === CHAR_SEPARATOR_TEMPLATE && pastedValue.length > counts.chars) {
      counts.separators = counts.separators + 1
    }
  }

  return counts
}

/**
 * comes value from input event and prepares to mask style and moves cursor to need position
 * example 123 -> 12/3
 */
export const getMaskValue = (
  {
    value,
    selectionStart,
    mask,
    separators,
    validators,
    prevMaskValue,
    cursorPosition,
    errorsCB,
    pastedValue,
  }: {
      value: string,
      mask: string,
      separators: TSeparators,
      validators: TValidator[],
      selectionStart: number,
      prevMaskValue: string,
      pastedValue: string,
      cursorPosition: number,
      errorsCB: (errors: string[]) => void,
    },
): { toMoveCursor: number, toInsertValue: string } | null => {
  // should prepared entered value to mask
  // for example to be __/__/__ mask
  // we entered 22| and would edition 223|
  // 223| we prepared to 22/3
  const {
    maskedValue,
    originalValue,
  } = prepareValueToMask(value, mask, separators)

  const errorsFromValidators: string[] = []

  const maxCursorPosition = mask.length || MIN_MAX_LENGTH

  const slicedMaskValue = maskedValue.slice(0, maxCursorPosition)

  let toMoveCursorAfterPasted = 0

  if (pastedValue) {
    const counts = getCharsInfo(
      mask,
      separators,
      cursorPosition,
      pastedValue,
    )

    toMoveCursorAfterPasted = counts.chars + counts.separators
  }

  const errorsFromCustomValidators = handleValidators(
    prepareValueFromMask(slicedMaskValue, separators),
  )(validators)

  if (errorsFromCustomValidators && errorsFromCustomValidators.length) {
    errorsFromValidators.push(...errorsFromCustomValidators)
  }

  if (originalValue.length && errorsFromValidators.length) {
    errorsCB(errorsFromValidators)

    return null
  }

  return getCursorPosition({
    nextMaskValue: slicedMaskValue,
    toMove: selectionStart,
    prevMaskValue,
    separators,
    cursorPosition,
    toMoveCursorAfterPasted,
    mask,
  })
}

export const makePasteHandler = (setPastedValue: (value: string) => void) =>
  (e: ClipboardEvent) => {
    // Get pasted data via clipboard API
    const clipboardData = e.clipboardData || (window as any).clipboardData
    const pastedData = clipboardData.getData("Text")

    setPastedValue(pastedData)
  }

export const joinClasses = (...props) => props.filter(Boolean).join(" ")

export const setCSSVar = (label: HTMLLabelElement, name: string, value: string) => {
  label.style.setProperty(name, value)
}
