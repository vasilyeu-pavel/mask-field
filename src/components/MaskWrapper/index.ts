import MaskWrapper from "./MaskWrapper"

export default MaskWrapper
