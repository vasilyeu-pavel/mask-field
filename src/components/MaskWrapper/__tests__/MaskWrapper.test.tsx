import React from "react"
import { mount, shallow } from "enzyme"
import { IMaskWrapperProps, typeCasting } from "../../../types"
import { setCSSVar } from "../../../utils"
import { MaskWrapper } from "../MaskWrapper"

jest.mock("../../../utils", () => {
  const originalModule = jest.requireActual("../../../utils")

  return {
    ...originalModule,
    setCSSVar: jest.fn(),
  }
})

jest.mock("react", () => {
  const originalModule = jest.requireActual("react")

  return {
    ...originalModule,
    useState: jest.fn(),
  }
})

Object.defineProperty(document, "fonts", {
  value: {
    ready: Promise.resolve(),
  },
  writable: true,
})

const mask = "YYYYMMDD-NNNN"

const defaultProps = typeCasting<IMaskWrapperProps>({
  mask,
  placeholder: "Personal number",
  value: "1",
  children: function fakeChildren () { return <input /> },
  separators: ["-"],
  onChange: jest.fn(),
  errors: null,
  modifiers: "mask",
  handleErrors: jest.fn(),
})

describe("MaskWrapper", () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  test("snapshot", () => {
    jest.spyOn(React, "useState")
      .mockImplementationOnce(() => ["YYYY", jest.fn()])
      .mockImplementationOnce(() => [false, jest.fn()])
      .mockImplementationOnce(() => [0, jest.fn()])
      .mockImplementationOnce(() => ["", jest.fn()])
      .mockImplementationOnce(() => ["", jest.fn()])

    const component = shallow(<MaskWrapper {...defaultProps} />)

    expect(component).toMatchSnapshot()
    expect(component.find(".focus").length).toBe(0)
  })

  test("should exist focus class", () => {
    jest.spyOn(React, "useState")
      .mockImplementationOnce(() => ["YYYY", jest.fn()])
      .mockImplementationOnce(() => [true, jest.fn()])
      .mockImplementationOnce(() => [0, jest.fn()])
      .mockImplementationOnce(() => ["", jest.fn()])
      .mockImplementationOnce(() => ["", jest.fn()])

    const component = shallow(<MaskWrapper {...defaultProps} />)

    expect(component.find(".focus").length).toBe(1)
  })

  test("should exist error class", () => {
    jest.spyOn(React, "useState")
      .mockImplementationOnce(() => ["YYYY", jest.fn()])
      .mockImplementationOnce(() => [true, jest.fn()])
      .mockImplementationOnce(() => [0, jest.fn()])
      .mockImplementationOnce(() => ["", jest.fn()])
      .mockImplementationOnce(() => ["", jest.fn()])

    const component = shallow(<MaskWrapper {...defaultProps} errors={["error"]} />)

    expect(component.find(".wrong").length).toBe(1)
  })

  test("should set maskValue in useEffect", () => {
    const value = "101"
    const offsetWidth = 100

    const setMaskValue = jest.fn()

    Object.defineProperty(HTMLElement.prototype, "offsetWidth", {
      configurable: true,
      value: offsetWidth,
    })

    jest.spyOn(React, "useState")
      .mockImplementationOnce(() => ["10", setMaskValue])
      .mockImplementationOnce(() => [true, jest.fn()])
      .mockImplementationOnce(() => [1, jest.fn()])
      .mockImplementationOnce(() => [value, jest.fn()])
      .mockImplementationOnce(() => ["", jest.fn()])

    jest.spyOn(React, "useRef").mockImplementationOnce(() => ({ current: null }))

    mount(<MaskWrapper {...defaultProps} value={value} />)

    expect(setMaskValue).toHaveBeenCalledWith("YMMDD-NNNN")
  })

  test("should set defaultMask in useEffect (!value)", () => {
    const value = ""
    const offsetWidth = 100

    const setMaskValue = jest.fn()

    jest.spyOn(React, "useState")
      .mockImplementationOnce(() => ["12", setMaskValue])
      .mockImplementationOnce(() => [true, jest.fn()])
      .mockImplementationOnce(() => [1, jest.fn()])
      .mockImplementationOnce(() => [value, jest.fn()])
      .mockImplementationOnce(() => ["", jest.fn()])

    Object.defineProperty(HTMLElement.prototype, "offsetWidth", {
      configurable: true,
      value: offsetWidth,
    })

    jest.spyOn(React, "useRef").mockImplementationOnce(() => ({ current: null }))

    mount(<MaskWrapper {...defaultProps} value={value} />)

    expect(setMaskValue).toHaveBeenCalledWith(mask)
  })

  test("should call setCssVar in useEffect", () => {
    const value = "101"
    const maskValue = "101"
    const offsetWidth = 100

    const setMaskValue = jest.fn()

    Object.defineProperty(HTMLElement.prototype, "offsetWidth", {
      configurable: true,
      value: offsetWidth,
    })

    jest.spyOn(React, "useState")
      .mockImplementationOnce(() => [maskValue, setMaskValue])
      .mockImplementationOnce(() => [true, jest.fn()])
      .mockImplementationOnce(() => [1, jest.fn()])
      .mockImplementationOnce(() => [value, jest.fn()])
      .mockImplementationOnce(() => ["", jest.fn()])

    mount(<MaskWrapper {...defaultProps} value={value} />)

    expect(setCSSVar).toBeCalledTimes(1)
  })

  test("should call setCssVar in useEffect 2 counts", async () => {
    const value = "101"
    const maskValue = "101"
    const offsetWidth = 100
    const countsCalled = 2

    const setMaskValue = jest.fn()

    Object.defineProperty(HTMLElement.prototype, "offsetWidth", {
      configurable: true,
      value: offsetWidth,
    })

    jest.spyOn(React, "useState")
      .mockImplementationOnce(() => [maskValue, setMaskValue])
      .mockImplementationOnce(() => [true, jest.fn()])
      .mockImplementationOnce(() => [1, jest.fn()])
      .mockImplementationOnce(() => [value, jest.fn()])
      .mockImplementationOnce(() => ["", jest.fn()])

    await mount(<MaskWrapper {...defaultProps} value={value} />)

    expect(setCSSVar).toBeCalledTimes(countsCalled)
  })

  test("should call moveCursorToEnd in useEffect", async () => {
    const value = "101"
    const offsetWidth = 100
    const cursorPosition = value.length
    const setCursorPosition = jest.fn()

    Object.defineProperty(HTMLElement.prototype, "offsetWidth", {
      configurable: true,
      value: offsetWidth,
    })

    jest.spyOn(React, "useState")
      .mockImplementationOnce(() => [value, jest.fn()])
      .mockImplementationOnce(() => [true, jest.fn()])
      .mockImplementationOnce(() => [1, setCursorPosition])
      .mockImplementationOnce(() => [value, jest.fn()])
      .mockImplementationOnce(() => ["", jest.fn()])

    mount(<MaskWrapper {...defaultProps} value={value} />)

    expect(setCursorPosition).toBeCalledWith(cursorPosition)
  })

  test("should call setValueToMaskStyle in useEffect", async () => {
    const value = "10112321333"
    const offsetWidth = 100
    const setValueToMaskStyle = jest.fn()

    Object.defineProperty(HTMLElement.prototype, "offsetWidth", {
      configurable: true,
      value: offsetWidth,
    })

    jest.spyOn(React, "useState")
      .mockImplementationOnce(() => [value, jest.fn()])
      .mockImplementationOnce(() => [true, jest.fn()])
      .mockImplementationOnce(() => [1, jest.fn()])
      .mockImplementationOnce(() => [value, setValueToMaskStyle])
      .mockImplementationOnce(() => ["", jest.fn()])

    mount(<MaskWrapper {...defaultProps} value={value} />)

    expect(setValueToMaskStyle).toBeCalledWith("10112321-333")
  })
})
