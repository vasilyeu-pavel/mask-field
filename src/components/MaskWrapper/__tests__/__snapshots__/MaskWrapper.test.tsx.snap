// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MaskWrapper snapshot 1`] = `
<label
  className="maskContainer mask mask"
  onClick={[Function]}
>
  <span
    className="fakeSpan mask__fake"
  />
  <input />
  <span
    className="mask mask__tmp"
    dangerouslySetInnerHTML={
      Object {
        "__html": "YYYY",
      }
    }
  />
</label>
`;
